### PR TITLE
support operate remote kubernetes cluster using tramp

### PR DESCRIPTION
**Pull request by RCmerci**
operating a remote cluster k8s using tramp,
e.g. when we are in a remote eshell(default-directory is
'/ssh:root@1.2.3.4:/') then input 'kubel <RET>',
we get a list of pods just as we input 'kubectl get po<RET>' in this
remote eshell.
also complete other functions(exec pod, edit resource, ...)

### CODE REVIEW

**Comment by abrochard** [comment]
Thank you for your interest! I'm a bit swamped these days and this is quite a bit of code, I will get to it for sure but if you could help me understand better some of the changes, it would be greatly appreciated. And just to confirm, this is to make `kubel` work when you are a remote host right?

File: kubel.el [hunk 1]
```diff
@@ -328,8 +328,8 @@ READONLY If true buffer will be in readonly mode(view-mode)."
   (when (get-buffer buffer-name)
     (kill-buffer buffer-name))
   (if async
-      (apply #'start-process buffer-name buffer-name "kubectl" (append (kubel--get-context-namespace) args))
-    (apply #'call-process "kubectl" nil buffer-name nil (append (kubel--get-context-namespace) args)))
+      (apply #'start-file-process buffer-name buffer-name "kubectl" (append (kubel--get-context-namespace) args))
```
Comment (RCmerci) [inline kubel.el:331]: use 'start-file-process' and 'process-file' to run a process in a remote directory

File: kubel.el [hunk 2]
```diff
@@ -442,13 +442,20 @@ Use C-c C-c to kubectl apply the current yaml buffer."
 (defun kubel-apply ()
   "Save the current buffer to a temp file and try to kubectl apply it."
   (interactive)
-  (let ((filename (format "/tmp/kubel/%s-%s.yaml"
-                          (replace-regexp-in-string "\*\\| " "" (buffer-name))
-                          (floor (float-time)))))
-    (unless  (file-exists-p "/tmp/kubel")
-      (make-directory "/tmp/kubel"))
+  (setq dir-prefix (or
```
Comment (RCmerci) [inline kubel.el:445]: add dir-prefix to filename to save tmp files on remote hosts when 'default-directory' is a tramp dir.

File: kubel.el [hunk 3]
```diff
@@ -532,26 +539,29 @@ ARGS is the arguments list from transient."
...
+                                     nil nil nil nil "default"))
+	 (kubel--buffer (get-buffer (kubel--buffer-name)))
+	 (last-default-directory (when kubel--buffer
+				   (with-current-buffer kubel--buffer default-directory))))
+    (when kubel--buffer (kill-buffer kubel--buffer))
     (setq kubel-namespace namespace)
     (unless (member namespace kubel-namespace-history)
       (push namespace kubel-namespace-history))
-    (kubel)))
+    (kubel last-default-directory)))
```
Comment (RCmerci) [inline kubel.el:551]: make sure the new created buffer in 'kubel' having same 'default-directory' with the killed buffer

File: kubel.el [hunk 4]
```diff
@@ -532,26 +539,29 @@ ARGS is the arguments list from transient."
...
-  (kubel))
+  (let* ((kubel--buffer (get-buffer (kubel--buffer-name)))
+	 (last-default-directory (when kubel--buffer (with-current-buffer kubel--buffer default-directory))))
+    (when kubel--buffer (kill-buffer kubel--buffer));; kill buffer for previous context if possible
+    (setq kubel-context
+          (completing-read
+           "Select context: "
+           (split-string (shell-command-to-string "kubectl config view -o jsonpath='{.contexts[*].name}'") " ")))
+    (kubel--invalidate-context-caches)
+    (kubel last-default-directory)))
```
Comment (RCmerci) [inline kubel.el:564]: same as above

File: kubel.el [hunk 5]
```diff
@@ -563,18 +573,19 @@ If called with a prefix argument REFRESH, refreshes
...
+	 (kubel--buffer (get-buffer current-buffer-name))
+	 (last-default-directory (when kubel--buffer (with-current-buffer kubel--buffer default-directory))))
     (setq kubel-resource
-	      (completing-read "Select resource: " resource-list))
-    (when (get-buffer current-buffer-name) ;; kill the current buffer to avoid confusion
-      (kill-buffer current-buffer-name)))
-  (kubel))
+	  (completing-read "Select resource: " resource-list))
+    (when kubel--buffer (kill-buffer kubel--buffer)) ;; kill buffer for previous context if possible
+    (kubel last-default-directory)))
```
Comment (RCmerci) [inline kubel.el:588]: same as above

File: kubel.el [hunk 6]
```diff
@@ -608,9 +619,14 @@ P is the port as integer."
...
                  (tramp-remote-shell-args  ("-i" "-c")))) ;; add the current context/namespace to tramp methods
-  (find-file (format "/kubectl:%s:/" (if (kubel--is-pod-view)
-                                         (kubel--get-resource-under-cursor)
-                                       (kubel--select-resource "Pods")))))
+  (setq dir-prefix (or
+		    (when (tramp-tramp-file-p default-directory)
+		      (with-parsed-tramp-file-name default-directory nil
+			(format "%s%s:%s@%s|" (or hop "") method user host)))
+		    ""))
+  (find-file (format "/%skubectl:%s:/" dir-prefix (if (kubel--is-pod-view)
```
Comment (RCmerci) [inline kubel.el:627]: concat tramp hops to kubectl exec the pods on remote cluster.

**Comment by RCmerci** [comment]
@abrochard , I have added some comments on files i changed.
And, yes, I tested kubel on my mac with k8s cluster on remote hosts.

**Review by abrochard — APPROVED** [review]
Awesome! Thank you for the annotations. It really helped quite a bit. And that change looks super cool.

**Pull request merged**